### PR TITLE
Elixir MQTT Client Macro

### PR DESCRIPTION
**Pull request by ericr3r**
Initial version of an Elixir MQTT client and a couple of tests for basic subscribe, publish and last will and testament.

The tests depend on an existing MQTT broker running locally

### CODE REVIEW

File: lib/mqtt/client.ex [hunk 1]
```diff
@@ -1,2 +1,165 @@
...
+      {:publish, topic :: String.t, message :: iodata, %{optional(:qos) => :mqtt_packet.qos, optional(:retain) => boolean}} |
+      {:reply, :gen_statem.from(), term()}
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour :mqtt_client
+
+      @doc false
+      def init(_) do
```
Comment (asabil) [inline lib/mqtt/client.ex]: I don't think the `init` callback is optional

File: lib/mqtt/client.ex [hunk 2]
```diff
@@ -1,2 +1,165 @@
...
+        {:ok, %{}}
+      end
+
+      @doc false
+      def handle_connect(_session_present, state) do
+          IO.puts("Macro handle connect")
```
Comment (asabil) [inline lib/mqtt/client.ex]: Could you please remove this IO.puts

File: lib/mqtt/client.ex [hunk 3]
```diff
@@ -1,2 +1,165 @@
...
+    end
+  end
+
+  @spec start_link(module(), any(), map()) :: {:ok, pid}
+  def start_link(module, args, options) do
+    :mqtt_client.start_link(module, args, convert_start_options(options))
+  end
+
+  @spec convert_start_options(start_options() | map()) :: start_options()
+  defp convert_start_options(options) do
```
Comment (asabil) [inline lib/mqtt/client.ex]: Maybe instead of doing this, we could change the underlying Erlang code to accept `iodata`for the host?

**Review by asabil — CHANGES_REQUESTED** [review]
Thank you for your contribution, this looks great and very much needed 👍 

I have left some comments, could you please have a look so we can get this landed.

**Comment by ericr3r** [comment]
@asabil, I'm happy to help. The library looks great so far, I'm glad to see an actively maintained  MQTT client. Please squash my commits before merging, I tend to commit often but don't want to pollute the project.

I'm fairly new to Elixir but I'm starting to use this library for a Nerves project. If you have time to review, I'll continue to add tests as I make changes. Also, I'm happy to add some additional docs before the next release and update the readme to reflect the limited tests. If you're comfortable with even some of the low level tests being written in Elixir, I can add some coverage there.

The areas that I'll look at improving are:
- Auto disconnect if missing PingResp
- Availability of explicit Disconnect action for clean disconnect

**Review by ericr3r — COMMENTED** [review]
@asabil Please review the changes I made and let me know if there is anything remaining or you want me to rebase and squash in a new pull request.

**Comment by asabil** [comment]
@erauer thanks again for the help and very sorry for the delayed answer.

Your help with erlang-mqtt is very much appreciated, and I would love to see some more documentation and tests go in before the next release. I think we can start by having the tests in Elixir if that is what you are most comfortable with, with the goal of hopefully moving these tests one layer below into Erlang land.

File: lib/mqtt/client.ex [hunk 4]
```diff
@@ -1,2 +1,157 @@
...
+      {:publish, topic :: String.t, message :: iodata, %{optional(:qos) => :mqtt_packet.qos, optional(:retain) => boolean}} |
+      {:reply, :gen_statem.from(), term()}
+
+  @doc false
+  defmacro __using__(_) do
+    quote location: :keep do
+      @behaviour :mqtt_client
+
+      @doc false
+      def init([]) do
```
Comment (asabil) [inline lib/mqtt/client.ex]: I think this default implementation should be completely removed. I can't think of any valid use case where you wouldn't want to implement the `init` callback

File: lib/mqtt/client.ex [hunk 5]
```diff
@@ -1,2 +1,157 @@
...
+          {:ok, state}
+      end
+
+      @doc false
+      def handle_call(msg, from, state) do
+        {:ok, state, {:reply, from, {:error, {:bad_call, msg}}}}
+      end
+
+      @doc false
+      def handle_info(msg, state) do
```
Comment (asabil) [inline lib/mqtt/client.ex]: Can we move the `handle_info` below the `handle_cast` ? This is the standard OTP callback order

File: lib/mqtt/client.ex [hunk 6]
```diff
@@ -1,2 +1,157 @@
...
+      def terminate(_reason, _state) do
+        :ok
+      end
+
+      @doc false
+      def code_change(_old, state, _extra) do
+        {:ok, state}
+      end
+
+      defoverridable [init: 1, handle_connect: 2, handle_connect_error: 2,
```
Comment (asabil) [inline lib/mqtt/client.ex]: Same as above, `init` should be removed from this list.

File: src/mqtt_client.erl [hunk 7]
```diff
@@ -148,7 +148,10 @@ start_options(Opts) ->
     }.
 
 start_options_transport(KeepAlive, {Type, Opts}) ->
-    Host = maps:get(host, Opts, "localhost"),
```
Comment (asabil) [inline src/mqtt_client.erl]: Actually I think this should be moved even below into `mqtt_transport`.

Today we have
`mqtt_transport:host() :: inet:ip_address() | inet:hostname().`

If we extend it to be
`mqtt_transport:host() :: inet:ip_address() | inet:hostname() | binary().`

Then we will just need to add a clause where we handle the binary case in `mqtt_transport:open/5`.

**Review by asabil — CHANGES_REQUESTED** [review]

File: src/mqtt_transport.erl [hunk 8]
```diff
@@ -44,6 +44,8 @@ open(Type, Host, Port, Options) ->
     open(Type, Host, Port, Options, infinity).
 
 -spec open(tcp | ssl, host(), port_number(), tcp_options() | ssl_options(), timeout()) -> {ok, transport()} | {error, term()}.
+open(Protocol, Host, Port, Options, Timeout) when is_binary(Host) ->
+    open_tcp(unicode:characters_to_list(Host), Port, Options, Timeout);
```
Comment (asabil) [inline src/mqtt_transport.erl]: I think this should be
`open(Protocol, unicode:characters_to_list(Host), Port, Options, Timeout);`

Comment (ericr3r) [reply src/mqtt_transport.erl]: good catch, this would have definitely broken SSL connections when hostname is binary. Updated in the pull request

**Review by asabil — APPROVED** [review]

**Comment by asabil** [comment]
Thank you very much @erauer and sorry it took longer than normal to get this merged.

**Pull request merged**